include/deal.II/matrix_free/hanging_nodes_internal.h

```diff
@@ -384,386 +521,249 @@ namespace internal
+
+              const bool flipped =
+                cell->line_orientation(line_no) !=
+                neighbor_cell->line_orientation(local_line_neighbor);
+
+              for (unsigned int base_element_index = 0, comp = 0;
+                   base_element_index < cell->get_fe().n_base_elements();
+                   ++base_element_index)
+                for (unsigned int c = 0;
+                     c <
+                     cell->get_fe().element_multiplicity(base_element_index);
+                     ++c, ++comp)
+                  {
+                    if (component_masks[cell->active_fe_index()][comp] == false)
+                      continue;
 
-            const auto &fe_base =
-              cell->get_fe().base_element(base_element_index);
+                    const unsigned int n_dofs_1d =
+                      cell->get_fe()
+                        .base_element(base_element_index)
+                        .tensor_degree() +
+                      1;
+
+                    for (unsigned int i = 0; i < n_dofs_1d; ++i)
+                      dof_indices[line_dof_idx(line_no, i, n_dofs_1d) +
+                                  idx_offset[comp]] = neighbor_dofs_all_temp
+                        [line_dof_idx(local_line_neighbor,
+                                      flipped ? (n_dofs_1d - 1 - i) : i,
+                                      n_dofs_1d) +
+                         idx_offset[comp]];
+                  }
+            }
+    }
 
-            if (dim == 1 ||
-                dynamic_cast<const FE_Q<dim> *>(&fe_base) == nullptr)
-              continue;
 
-            const unsigned int fe_degree = fe_base.tensor_degree();
-            const unsigned int n_dofs_1d = fe_degree + 1;
-            const unsigned int dofs_per_face =
-              Utilities::fixed_power<dim - 1>(n_dofs_1d);
 
-            std::vector<types::global_dof_index> neighbor_dofs_all(
-              idx_offset.back());
-            std::vector<types::global_dof_index> neighbor_dofs_all_temp(
-              idx_offset.back());
+    template <int dim>
+    template <typename CellIterator>
+    inline bool
+    HangingNodes<dim>::setup_constraints(
+      const CellIterator &                                      cell,
+      const std::shared_ptr<const Utilities::MPI::Partitioner> &partitioner,
+      const std::vector<unsigned int> &     lexicographic_mapping,
+      std::vector<types::global_dof_index> &dof_indices,
+      const ArrayView<ConstraintKinds> &    masks) const
+    {
+      // 1) check if finite elements support fast hanging-node algorithm
+      const auto component_masks =
+        compute_component_mask(cell->get_dof_handler().get_fe_collection());
+
+      if ([](const auto &outer) {
+            for (const auto &inner : outer)
+              for (const auto i : inner)
+                if (i)
+                  return true;
+            return false;
+          }(component_masks) == false)
+        return false;
```

Comment:
I think this is hard to follow, why not simply
```suggestion
      return std::none_of(component_mask.begin(), component_mask.end(),
          [] (const auto &a) { return *std::max_element(a.begin(), a.end());});
```
Or write out the loops to find an entry with true (at the expense of an additional variable as we can't return from within the loop.